src/main/java/org/openjdk/jextract/impl/HeaderFileBuilder.java

```diff
@@ -167,33 +166,42 @@ class Holder {
                         if (TRACE_DOWNCALLS) {
                             traceDowncall(\{traceArgList});
                         }
-                        \{returnExpr}mh$.invokeExact(\{paramList});
+                        \{returnWithCast}mh$.invokeExact(\{paramList});
                     } catch (Throwable ex$) {
                        throw new AssertionError("should not reach here", ex$);
                     }
                 }
                 """);
         } else {
-            String invokerName = javaName + "$invoker";
-            String invokerFactoryName = javaName + "$makeInvoker";
             String paramExprs = paramExprs(declType, finalParamNames, isVarArg);
+            String varargsParam = finalParamNames.get(finalParamNames.size() - 1);
+            appendBlankLine();
+            emitDocComment(decl, "Variadic invoker interface for:");
             appendLines(STR."""
-                public interface \{invokerName} {
-                    \{retType} \{javaName}(\{paramExprs});
+                public interface \{javaName} {
+                    \{retType} apply(\{paramExprs});
+
+                    /**
+                     * Invoke the variadic function with the given parameters. Layouts for variadic arguments are inferred.
+                     */
+                    static \{retType} invoke(\{paramExprs}) {
+                        MemoryLayout[] inferredLayouts$ = \{runtimeHelperName()}.inferVariadicLayouts(\{varargsParam});
+                        \{returnNoCast}\{javaName}(inferredLayouts$).apply(\{String.join(", ", finalParamNames)});
+                    }
                 }
 
                 """);
             emitDocComment(decl);
```

Comment:
Maybe we could add a line to the doc comment explaining that this is a variadic invoker factory?